src/Psl/Option/Option.php

```diff
@@ -305,4 +305,77 @@ public function equals(mixed $other): bool
     {
         return Comparison\equal($this, $other);
     }
+
+    /**
+     * Combines two `Option` values into a single `Option` containing a tuple of the two inner values.
+     * If either of the `Option`s is `None`, the resulting `Option` will also be `None`.
+     *
+     * @note: If an element is `None`, the corresponding element in the resulting tuple will be `None`.
+     *
+     * @template U
+     *
+     * @param Option<U> $other The other `Option` to zip with.
+     *
+     * @return Option<array{T, U}> The resulting `Option` containing the combined tuple or `None`.
+     */
+    public function zip(Option $other): Option
+    {
+        return $this->andThen(static function ($a) use ($other) {
+            return $other->map(static fn($b) => [$a, $b]);
+        });
+    }
+
+    /**
+     * Applies the provided closure to the value contained in this `Option` and the value contained in the $other `Option`,
+     * and returns a new `Option` containing the result of the closure.
+     *
+     * @template U
+     * @template Tv
+     *
+     * @param Option<U> $other The Option to zip with.
+     * @param (Closure(T, U): Tv) $closure The closure to apply to the values.
+     *
+     * @return Option<Tv> The new `Option` containing the result of applying the closure to the values,
+     *                    or `None` if either this or the $other `Option is `None`.
+     */
+    public function zipWith(Option $other, Closure $closure): Option
+    {
+        return $this->andThen(
+            /** @param T $a */
+            static function ($a) use ($other, $closure) {
+                return $other->map(
+                    /** @param U $b */
+                    static fn ($b) => $closure($a, $b)
+                );
+            }
+        );
+    }
+
+    /**
+     * @template OValue1
+     * @template OValue2
+     *
+     * @psalm-if-this-is Option<array{OValue1, OValue2}>
+     *
+     * @return array{Option<OValue1>, Option<OValue2>}
+     */
+    public function unzip(): array
+    {
+        if ($this->option === null) {
+            return [none(), none()];
+        }
+
+        /** @psalm-suppress DocblockTypeContradiction we want this check in runtime */
```

Comment:
Maybe better to remove these 2 ifs and rather assert the type this way:

```php
Type\shape([Type\mixed(), Type\mixed()])->assert($this->option);
```

It makes more sense to me to trigger a type's Assertion error than a pair of none's in this case.

In PSL v2+, we are going for better static analyzer errors meaning we will deal less with runtime errors. However, in this case I suppose it makes a lot of sense to make sure that the given value can be unzipped.

Since phpstan does not have a equivalence for `psalm-if-this-is` ATM, it must be added to https://github.com/php-standard-library/phpstan-extension. However I don't see that as a blocking issue at the moment. Rather something that needs to be done for phpstan to infer the types correctly.


Can you provide a static analysis test-case in this folder as well to make sure it works (and keeps on working) as expected:
tests/static-analysis/Option

Reply:
This seems to work psalm-wise:

https://psalm.dev/r/9c6fa21644

I must admit it's a bit "hacky", but it's an error that shouldn't happen in runtime if you have your SAT doing the work for you. So I'm not sure here if it's a good idea to introduce an exception that should never occur cause it's being caught up front. WDYT?

Reply:
That's interesting! The other way I see also is to keep the "manual" type checking that I have already implemented and throw an `invariant_violation()`. We would miss the structure of the type error in the exception message.

Both are doable, one requiring a private method to perform that smart but hacky assertion that would give a complete `AssertionException`, and the other one would not add a private method and would throw an `InvariantViolationException`. Do you have any preference?

Reply:
> one requiring a private method to perform that smart but hacky assertion

You don't need a private function for that. You could do `Type\vec(Type\shape([Type\mixed(), Type\mixed()]))->assert($options);` 

Because that is what the `assertTuppleOption` is basically doing. (I've added it there to see how psalm behaves - but it is essentially the same). So I guess that would be the preferred solution for me ATM. Hope it won't bite us when a psalm update is performed though :)

Reply:
About this
> You don't need a private function for that. You could do Type\vec(Type\shape([Type\mixed(), Type\mixed()]))->assert($options);

There's this problem I mentioned above. Probably it's behaving differently than a normal psalm assertion because of the Psalm plugin, but if I'm not wrong just erases the types of the tuple inside the function

> >Maybe better to remove these 2 ifs and rather assert the type this way:
> >
> >```php
> >Type\shape([Type\mixed(), Type\mixed()])->assert($this->option);
> >```
> 
> There is a problem with this approach: it gives a false `PossiblyNullArrayAccess` after converting `list{list{OValue1:fn-psl\option\option::unzip as mixed, OValue2:fn-psl\option\option::unzip as mixed}}` (inferred by `@psalm-if-this-is`) back into `list{array{0: mixed, 1: mixed}}|null`, and is also assigning `mixed` to `$a` and `$b`.
> 
> While I was doing different cases for the tests, I tested and discarded a version in which I would manually verify the tuple, and throw a `Psl\Option\Exception\UnzipException`. Would you be OK with that, @veewee?